### DISCUSSION TITLE
Getting API tokens - how will it work?

### DISCUSSION THREAD

**Discussion by jbanyer** [amberelectric/public-api#27] [General]
Hi,

Currently an API token is obtained from https://app.amber.com.au/developers/, which is an authenticated page (and currently only available to selected beta tester users, I gather).

Will it work this way for all users?

Will there be an API endpoint or OAuth option for obtaining an API token?

Cheers,
Jay.

**Comment by madpilot** [comment]
Currently the API is only available to customers. We'll be adding demo accounts in the next few months and Oauth is on the roadmap, no ETA though (it'll be after demo accounts - because you'll need to login to set things up)

**Reply by jbanyer** [reply]
Thanks for your reply. To give context to my question, imagine you are writing a 3rd party home energy management system (HEMS) that integrates with the Amber API, and is used by many people (eg openHAB or similar).

For the end user to enable this integration, would you instruct the end user to obtain a token from https://app.amber.com.au/developers/, then copy and paste it into the HEMS config? Or could the HEMS system use OAuth or similar to obtain the token on the user's behalf, avoiding the need for them to copy-and-paste it?

I understand the API is only for customers, and I'm assuming the end user in this example is an Amber customer.

Cheers

**Reply by madpilot** [reply]
For systems that can reliably store secrets they would definitely be able to use OAuth (once implemented) to allow customers to authorise access to their account, however for open source projects like OpenHAB and HomeAssistant that isn't really viable for a couple of reasons:

1. They would need to expose their server to the outside world to complete the OAuth handshake (unless they used out of band auth which is an order of magnitude harder to do that generating an API key)
2. Each customer would need to generate app keys for OAuth which is, again harder than the current system

For closed third party apps, OAuth will definitely be an option though.

**Reply by madpilot** [reply]
I don't see us adding a token endpoint. I don't want to encourage users to enter their password on third party systems

**Reply by madpilot** [reply]
So to answer your actual question: yes, that would be the flow for OpenHAB or Home Assistant.

And that's the flow I'm implementing in Home Assistant: https://github.com/madpilot/hass-amber-electric
